servers/services/src/main/java/org/projectnessie/services/impl/TreeApiImplWithAuthorization.java

```diff
@@ -139,97 +146,82 @@ protected Hash deleteReference(NamedRef ref, String expectedHash)
-        responseCheck.check().keySet().stream().map(Check::contentId).collect(Collectors.toSet());
-
-    // Second, only add the `Entry`s for allowed content keys.
-
-    return entriesList.stream()
-        .filter(entry -> !notAllowedContentIds.contains(entry.getContentId()));
+    return super.filterEntries(refWithHash, filter)
+        .and(
+            entry ->
+                startAccessCheck()
+                    .canReadContentKey(
+                        refWithHash.getValue(), fromKey(entry.getKey()), entry.getContentId())
+                    .check()
+                    .isEmpty());
   }
 
   @Override
-  protected LogResponse getCommitLog(
-      Integer maxRecords,
+  protected <R> R getCommitLog(
       FetchOption fetchOption,
       String filter,
       WithHash<NamedRef> endRef,
-      String startHash)
+      String startHash,
+      PagedResponseHandler<R, LogEntry> pagedResponseHandler)
       throws NessieNotFoundException {
     NamedRef ref = endRef.getValue();
-
     startAccessCheck().canListCommitLog(ref).checkAndThrow();
-    LogResponse logResponse =
-        super.getCommitLog(maxRecords, fetchOption, filter, endRef, startHash);
+    return super.getCommitLog(
+        fetchOption,
+        filter,
+        endRef,
+        startHash,
+        new PagedResponseHandler<R, LogEntry>() {
 
-    Stream<Operation> allOperations =
-        logResponse.getLogEntries().stream()
-            .map(LogEntry::getOperations)
-            .filter(Objects::nonNull)
-            .flatMap(Collection::stream);
+          @Override
+          public R build() {
+            return pagedResponseHandler.build();
+          }
 
-    // Post-process the response to only contain content-keys to which the user has access to.
-    // Note: no optimization here to return the non-post-processed response to ensure that tests
-    // catch potential issues during post-processing early (e.g. a new attribute added to
-    // LogResponse, but not added here).
+          @Override
+          public boolean addEntry(LogEntry entry) {
+            if (entry.getOperations() != null) {
+              BatchAccessChecker responseCheck = startAccessCheck();
+              entry
+                  .getOperations()
+                  .forEach(
+                      op -> {
+                        if (op instanceof Put) {
+                          Put put = (Put) op;
+                          responseCheck.canReadContentKey(
+                              ref, put.getKey(), put.getContent().getId());
+                        } else if (op instanceof Delete) {
+                          Delete delete = (Delete) op;
+                          responseCheck.canReadContentKey(ref, delete.getKey(), null);
+                        }
+                      });
 
-    // First, collect the content-keys to which the user has no access to.
+              Set<ContentKey> notAllowed =
+                  responseCheck.check().keySet().stream()
```

Comment:
I think keeping the `maxRecords` parameter can be helpful. With it we could get a full uncensored page from storage then do all access checks in one batch.

We'd have to support getting a page token from an arbitrary element of the result set, though. I think technically it should be possible, because our paging tokens are likely going to be derived either from commit hashes or element names (or both), which are all available from the elements themselves (unless I'm missing something).

In that regard, we may have to change `PaginationIterator` to something that holds one page of results and provides continuation tokens for any requested element.

Granted, we may end up with an empty page in the end, but I think it is ok from the API POV. Also, it might be possible to re-request from upstream if the page ends up being empty.

Reply:
Actually, passing `maxResults` through and filtering might be enough (i.e. we can keep current end-of-page token approach).

I think it should be possible to make a utility class implementing `PagedResponseHandler` with batched filtering and making 1 or more calls to a lower-level API to fill the request to the max if intermediate entries end up being filtered out. WDYT?

Reply:
#5925 for follow-up